v3/src/transports/http/transport.ts

```diff
@@ -0,0 +1,214 @@
+   * the incoming request by the HTTP handler, which itself should invoke this
+   * function.
+   */
+  args: IProcessHttpRequestArgs,
+): Promise<IHttpResponse> {
+  const { schema, request } = args;
+
+  if (request.method !== 'POST' && request.method !== 'GET') {
+    return generatedResponse({
+      response: {
+        // TODO(AS3) This should be a HTTP status code 405.
+        errors: [new GraphQLError("Unsupported method")],
+      },
+    });
+  }
+
+  try {
+    return generatedResponse({
+      response: await processGraphqlRequest({
+        schema,
+        request: request.parsedRequest,
+      }),
+    });
+  } catch (err) {
+    return generatedResponse({
+      response: {
+        errors: [new GraphQLError("Internal server error")],
+      },
+    });
+  }
+}
+
+/**
+ * Generate the response to return to the HTTP transport.
+ *
+ * TODO(AS3) Yet to be determined, is how the setting of additional headers
+ * should be handled.
+ *
+ * @param response
+ */
+export function generatedResponse(args: {
+  response: GraphQLResponse;
+  headers?: OutgoingHttpHeaders;
+}): IHttpResponse {
+  const {
+    response,
+    headers = Object.create(null),
+  } = args;
+
+  /**
+   * In the future, GraphQL execution should return an `AsyncIterable`. However,
+   * today it returns a `Promise`, so we'll coerce it into an `AsyncIterable`
+   * with a generator function implemented on a `Symbol.asyncIterator` property.
+   */
+  const body = {
+    [Symbol.asyncIterator]: async function*() {
+      yield response;
+    },
+  };
+
+  const statusCode: number = statusCodeForResponse(response);
+
+  return {
+    body,
+    statusCode,
+    headers: {
+      "content-type": "application/json",
+      ...headers,
+    },
+  };
+}
+
+/**
+ * Convert a singular `GraphQLError` to an HTTP status code.
+ *
+ * TODO(AS3) This could be more defensive since, at the moment, we don't check
+ * this against a known error type — e.g. `ValidationError`, `ParseError` — as
+ * we did in Apollo Server 2.x.  For now, we just structurally check the error
+ * to see if it matches characteristics of particular error types.
+ *
+ * More , the new execution pipeline doesn't actually throw these errors
+ * right now, so this is the only technique we can currently employ.  While we
+ * could re-use `apollo-server-errors` errors in Apollo Server 3, we may want to
+ * consider another option, so I didn't want to reach for that package just yet.
+ *
+ * @param error A `GraphQLError` instance to reduce to a status code.
+ */
+export function statusCodeForError(error: GraphQLError): number {
+  if (!(error instanceof GraphQLError)) {
+    return 500;
+  }
+
+  if (
+    error.source &&
+    Array.isArray(error.positions) &&
+    error.positions.length &&
+    Array.isArray(error.locations) &&
+    error.locations.length
+  ) {
+    if (Array.isArray(error.nodes) && error.nodes.length) {
+      return 400; // Validation error
```

Comment:
We need to introduce specific types into the stages of the request pipeline to capture these precisely!

Reply:
Not yet.  We previously had `ValidationError` and `ParseError`, but those have not re-surfaced in the new machinery yet.  cc @trevor-scheer something we'll need.

Reply:
Sorry I didn't respond to this sooner!

> What is that breaking change? I thought we were just going to include `code` properties on errors, but I think we could still support `ApolloError` style errors, perhaps?

Right now, properties set on the `extensions` object are added as properties to the `Error` itself. This is old behavior that we chose to preserve but it's not spec-compliant. I wrote some code awhile back that gives us (and users) a migration path away from depending on this behavior with the intention that we'd break support for it in the future.
https://github.com/apollographql/apollo-server/blob/faba52c689c22472a19fcb65d78925df549077f7/packages/apollo-server-errors/src/index.ts#L40